coding.go

```diff
@@ -117,44 +95,42 @@ func (n *ProtoNode) GetPBNode() *pb.PBNode {
 		}
 	}
 
-	if len(n.data) > 0 {
-		pbn.Data = n.data
+	data := n.Data()
+	if len(data) > 0 {
+		pbn.Data = data
 	}
 	return pbn
 }
 
-// EncodeProtobuf returns the encoded raw data version of a Node instance.
-// It may use a cached encoded version, unless the force flag is given.
-func (n *ProtoNode) EncodeProtobuf(force bool) ([]byte, error) {
-	sort.Stable(LinkSlice(n.links)) // keep links sorted
-	if n.encoded == nil || force {
-		n.cached = cid.Undef
-		var err error
-		n.encoded, err = n.marshalImmutable()
-		if err != nil {
-			return nil, err
-		}
-	}
-
-	if !n.cached.Defined() {
-		c, err := n.CidBuilder().Sum(n.encoded.encoded)
-		if err != nil {
-			return nil, err
-		}
+type UnhashedProtoNode struct {
+	encoded []byte
+	node    dagpb.PBNode
+}
 
-		n.cached = c
+func (upn *UnhashedProtoNode) ToProtoNode(builder cid.Builder) (*ProtoNode, error) {
+	c, err := builder.Sum(upn.encoded)
+	if err != nil {
+		return nil, err
 	}
-
-	return n.encoded.encoded, nil
+	blk, err := blocks.NewBlockWithCid(upn.encoded, c)
+	if err != nil {
+		return nil, err
+	}
+	return &ProtoNode{blk, upn.node}, nil
 }
 
 // DecodeProtobuf decodes raw data and returns a new Node instance.
-func DecodeProtobuf(encoded []byte) (*ProtoNode, error) {
-	n, err := unmarshal(encoded)
+func DecodeProtobuf(encoded []byte) (*UnhashedProtoNode, error) {
```

Comment:
where are we calling this where we don't already have the CID associated with the data?

Reply:
yea, so partly the scary part is this is a public function, so the answer is "who knows". A cursory search using Github's organization wide search for IPFS shows several potential places.